core/src/main/java/org/phoenixframework/liveview/lib/Document.kt

```diff
@@ -0,0 +1,153 @@
+package org.phoenixframework.liveview.lib
+
+
+class Document {
+    private var nativeObject: Long
+
+
+    constructor() {
+        nativeObject = empty()
+    }
+
+    internal constructor(pointer: Long) {
+        nativeObject = pointer
+    }
+
+    companion object {
+        /** Parses a `Document` from a string
+         * @throws Exception if the document is malformed
+         * */
+        @Throws
+        fun parse(string: String): Document {
+            val result = JavaResult(do_parse(string))
+            return result.document ?: throw Exception(result.error)
+        }
+
+        private external fun do_parse(text: String): Long
+
+        /** Output logs from the Rust side into android's logcat */
+        private external fun initialize_log()
+        enum class ChangeType {
+            Change, Add, Remove
+        }
+
+        open class Handler {
+            private fun mOnHandle(
+                context: Long,
+                changeType: Byte,
+                nodeRef: Int,
+                parent: Int
+            ) {
+                onHandle(
+                    Document(context),
+                    ChangeType.values()[changeType.toInt()],
+                    NodeRef(nodeRef),
+                    if (parent == 0) null else NodeRef(parent)
+                )
+            }
+
+            open fun onHandle(
+                context: Document,
+                changeType: ChangeType,
+                nodeRef: NodeRef,
+                parent: NodeRef?
+            ){}
+        }
+
+    }
+
+    fun getNodeString(nodeRef: NodeRef): String {
+        return node_to_string(nativeObject, nodeRef.ref)
+    }
+
+    /** Returns the root node of the document
+     * The root node can be used in insertion operations, but can not have attributes applied to it
+     */
+    // could also use lazy
+    val rootNodeRef get() = run { NodeRef(root(nativeObject)) }
+
+    /** Returns the data associated with the given `NodeRef` */
+    fun getNode(nodeRef: NodeRef): Node {
+        val nodePtr = get_node(nativeObject, nodeRef.ref)
+        // construct node
+        return when (get_node_type(nodePtr)) {
+            0.toByte() -> {
+                Node.Root
+            }
+            1.toByte() -> {
+                // element
+                val elementPtr = get_node_element(nodePtr)
+                val namespace = get_node_element_namespace(elementPtr)
+                val tag = get_node_element_tag(elementPtr)
+                val attributes = get_node_element_attributes(elementPtr)
+                Node.Element(namespace, tag, attributes).also { it.nativeObject = elementPtr }
+            }
+            2.toByte() -> {
+                Node.Leaf(get_node_leaf_string(nativeObject, nodeRef.ref))
+            }
+            else -> throw Exception("Unreachable code")
+        }
```

Comment:
Include the node type in the error in case a node type is added later and this code is not updated

```suggestion
        return when (val nodeType = get_node_type(nodePtr)) {
            0.toByte() -> {
                Node.Root
            }
            1.toByte() -> {
                // element
                val elementPtr = get_node_element(nodePtr)
                val namespace = get_node_element_namespace(elementPtr)
                val tag = get_node_element_tag(elementPtr)
                val attributes = get_node_element_attributes(elementPtr)
                Node.Element(namespace, tag, attributes).also { it.nativeObject = elementPtr }
            }
            2.toByte() -> {
                Node.Leaf(get_node_leaf_string(nativeObject, nodeRef.ref))
            }
            else -> throw Exception("Unknown node type ${nodeType}")
        }
```